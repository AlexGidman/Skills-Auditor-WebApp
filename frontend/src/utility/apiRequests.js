import axios from "axios";

const categoryEndpoint = "/api/category";
const skillEndpoint = "/api/skill";
const userEndpoint = "/api/user";
const directReportEndpoint = "/api/directreport";
const loginEndpoint = "/api/login";
const userSkillEndpoint = "/api/userskill";

//CATEGORY
export const getAllCategories = async (authHeaders) => {
    return axios.get(categoryEndpoint, authHeaders);
};

export const getCategory = async (authHeaders, id) => {
    return axios.get(`${categoryEndpoint}/${id}`, authHeaders);
};

export const addCategory = async (authHeaders, name) => {
    const data = {
        name,
    };
    return axios.post(`${categoryEndpoint}`, data, authHeaders);
};

export const updateCategory = async (authHeaders, id, name) => {
    const data = {
        id,
        name,
    };
    return axios.put(categoryEndpoint, data, authHeaders);
};

export const deleteCategory = async (authHeaders, id) => {
    const data = { id };
    return axios.delete(categoryEndpoint, { ...authHeaders, data });
};

//SKILL
export const getAllSkills = async (authHeaders) => {
    return axios.get(skillEndpoint, authHeaders);
};

export const getSkill = async (authHeaders, id) => {
    return axios.get(`${skillEndpoint}/${id}`, authHeaders);
};

export const addSkill = async (authHeaders, name, category_id) => {
    const data = {
        name,
        category_id,
    };
    return axios.post(`${skillEndpoint}`, data, authHeaders);
};

export const updateSkill = async (authHeaders, id, name, categoryId) => {
    const data = {
        id,
        name,
        category_id: categoryId,
    };
    return axios.put(skillEndpoint, data, authHeaders);
};

export const deleteSkill = async (authHeaders, id) => {
    const data = { id };
    return axios.delete(skillEndpoint, { ...authHeaders, data });
};

//USER
export const getUserDetails = async (authHeaders, id) => {
    return axios.get(`${userEndpoint}/${id}`, authHeaders);
};

export const getUserFromToken = async (authHeaders) => {
    return axios.get(`${userEndpoint}/token`, authHeaders);
};

export const updateUserDetails = async (
    authHeaders,
    id,
    forename,
    surname,
    email,
    password,
    job_role,
    system_role,
) => {
    const data = {
        id,
        forename,
        surname,
        email,
        password,
        job_role,
        system_role,
    };

    return axios.put(`${userEndpoint}`, data, authHeaders);
};

export const addUser = async (
    authHeaders,
    forename,
    surname,
    email,
    password,
    job_role,
    system_role,
) => {
    const data = {
        forename,
        surname,
        email,
        password,
        job_role,
        system_role,
    };
    return axios.post(userEndpoint, data, authHeaders);
};

export const deleteUser = async (authHeaders, id) => {
    const data = { id };
    return axios.delete(userEndpoint, { ...authHeaders, data });
};

//DIRECT REPORTS
export const getAllDirectReports = async (authHeaders, id) => {
    return axios.get(`${directReportEndpoint}/${id}`, authHeaders);
};

export const addDirectReport = async (authHeaders, userId, reportId) => {
    const data = {
        user_id: userId,
        report_id: reportId,
    };
    return axios.post(directReportEndpoint, data, authHeaders);
};

//STAFF SKILLS
export const getAllStaffSkills = async (authHeaders, id) => {
    return axios.get(`${userSkillEndpoint}/byuser/${id}`, authHeaders);
};

export const getStaffSkill = async (authHeaders, id) => {
    return axios.get(`${userSkillEndpoint}/${id}`, authHeaders);
};

export const deleteStaffSkill = async (authHeaders, id) => {
    const data = { id };
    return axios.delete(userSkillEndpoint, { ...authHeaders, data });
};

export const addUserSkill = async (
    authHeaders,
    user_id,
    skill_id,
    skill_level,
    notes,
    expiry_date,
) => {
    const data = {
        user_id,
        skill_id,
        skill_level,
        notes,
        expiry_date,
    };
    return axios.post(`${userSkillEndpoint}`, data, authHeaders);
};

export const updateStaffSkill = async (authHeaders, id, skill_level, notes) => {
    const data = {
        id,
        skill_level,
        notes,
    };

    return axios.put(`${userSkillEndpoint}`, data, authHeaders);
};

//LOGIN

export const postLogin = async (email, password) => {
    const credentials = { email, password };
    return axios.post(`${loginEndpoint}`, credentials);
};

export const validateToken = async (authHeaders) => {
    return axios.get(`${loginEndpoint}/validate`, authHeaders);
};
