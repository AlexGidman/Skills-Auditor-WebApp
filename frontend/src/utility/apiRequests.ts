import axios, { AxiosRequestHeaders } from "axios";

const categoryEndpoint = "/api/category";
const skillEndpoint = "/api/skill";
const userEndpoint = "/api/user";
const directReportEndpoint = "/api/directreport";
const loginEndpoint = "/api/login";
const userSkillEndpoint = "/api/userskill";

//CATEGORY
export const getAllCategories = async (authHeaders: AxiosRequestHeaders) => {
    return axios.get(categoryEndpoint, authHeaders);
};

export const getCategory = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${categoryEndpoint}/${id}`, authHeaders);
};

export const addCategory = async (authHeaders: AxiosRequestHeaders, name: string) => {
    const data = {
        name,
    };
    return axios.post(`${categoryEndpoint}`, data, authHeaders);
};

export const updateCategory = async (
    authHeaders: AxiosRequestHeaders,
    id: string,
    name: string,
) => {
    const data = {
        id,
        name,
    };
    return axios.put(categoryEndpoint, data, authHeaders);
};

export const deleteCategory = async (authHeaders: AxiosRequestHeaders, id: string) => {
    const data = { id };
    return axios.delete(categoryEndpoint, { ...authHeaders, data });
};

//SKILL
export const getAllSkills = async (authHeaders: AxiosRequestHeaders) => {
    return axios.get(skillEndpoint, authHeaders);
};

export const getSkill = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${skillEndpoint}/${id}`, authHeaders);
};

export const addSkill = async (
    authHeaders: AxiosRequestHeaders,
    name: string,
    categoryId: string,
) => {
    const data = {
        name,
        categoryId,
    };
    return axios.post(`${skillEndpoint}`, data, authHeaders);
};

export const updateSkill = async (
    authHeaders: AxiosRequestHeaders,
    id: string,
    name: string,
    categoryId: string,
) => {
    const data = {
        id,
        name,
        categoryId: categoryId,
    };
    return axios.put(skillEndpoint, data, authHeaders);
};

export const deleteSkill = async (authHeaders: AxiosRequestHeaders, id: string) => {
    const data = { id };
    return axios.delete(skillEndpoint, { ...authHeaders, data });
};

//USER
export const getUserDetails = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${userEndpoint}/${id}`, authHeaders);
};

export const getUserFromToken = async (authHeaders: AxiosRequestHeaders) => {
    return axios.get(`${userEndpoint}/token`, authHeaders);
};

export const updateUserDetails = async (
    authHeaders: AxiosRequestHeaders,
    id: string,
    firstName: string,
    lastName: string,
    email: string,
    password: string,
    jobRole: string, // TODO: change when job role type changes
    systemRole: string, // TODO: change when job system type changes
) => {
    const data = {
        id,
        firstName,
        lastName,
        email,
        password,
        jobRole,
        systemRole,
    };

    return axios.put(`${userEndpoint}`, data, authHeaders);
};

export const addUser = async (
    authHeaders: AxiosRequestHeaders,
    firstName: string,
    lastName: string,
    email: string,
    password: string,
    jobRole: string, // TODO: change when job role type changes
    systemRole: string, // TODO: change when job system type changes
) => {
    const data = {
        firstName,
        lastName,
        email,
        password,
        jobRole,
        systemRole,
    };
    return axios.post(userEndpoint, data, authHeaders);
};

export const deleteUser = async (authHeaders: AxiosRequestHeaders, id: string) => {
    const data = { id };
    return axios.delete(userEndpoint, { ...authHeaders, data });
};

//DIRECT REPORTS
export const getAllDirectReports = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${directReportEndpoint}/${id}`, authHeaders);
};

export const addDirectReport = async (
    authHeaders: AxiosRequestHeaders,
    userId: string,
    reportId: string,
) => {
    const data = {
        userId: userId,
        reportId: reportId,
    };
    return axios.post(directReportEndpoint, data, authHeaders);
};

//STAFF SKILLS
export const getAllStaffSkills = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${userSkillEndpoint}/byuser/${id}`, authHeaders);
};

export const getStaffSkill = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${userSkillEndpoint}/${id}`, authHeaders);
};

export const deleteStaffSkill = async (authHeaders: AxiosRequestHeaders, id: string) => {
    const data = { id };
    return axios.delete(userSkillEndpoint, { ...authHeaders, data });
};

export const addUserSkill = async (
    authHeaders: AxiosRequestHeaders,
    userId: string,
    skillId: string,
    skillLevel: number, // TODO: chnage this when skill level type changes
    notes: string,
    expiryDate: Date,
) => {
    const data = {
        userId,
        skillId,
        skillLevel,
        notes,
        expiryDate,
    };
    return axios.post(`${userSkillEndpoint}`, data, authHeaders);
};

export const updateStaffSkill = async (
    authHeaders: AxiosRequestHeaders,
    id: string,
    skillLevel: number, // TODO: chnage this when skill level type changes
    notes: string,
) => {
    const data = {
        id,
        skillLevel,
        notes,
    };

    return axios.put(`${userSkillEndpoint}`, data, authHeaders);
};

//LOGIN

export const postLogin = async (email: string, password: string) => {
    const credentials = { email, password };
    return axios.post(`${loginEndpoint}`, credentials);
};

export const validateToken = async (authHeaders: AxiosRequestHeaders) => {
    return axios.get(`${loginEndpoint}/validate`, authHeaders);
};
