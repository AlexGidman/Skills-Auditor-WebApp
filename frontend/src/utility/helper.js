import { useEffect, useState } from "react";
import { useOutletContext, useNavigate } from "react-router-dom";
import {
    ADMIN,
    BASIC,
    EXPERT,
    HIGH,
    NONE,
    INTERMEDIATE,
    MANAGER_SR,
    SYSTEM_ROLES,
    TOKEN_COOKIE,
} from "./types";
import { useCookies } from "react-cookie";

export const useLogout = () => {
    const [, , removeCookie] = useCookies(TOKEN_COOKIE);
    const navigate = useNavigate();
    const [logout, setLogout] = useState(false);

    useEffect(() => {
        if (logout) {
            setLogout(false);
            removeCookie(TOKEN_COOKIE);
            navigate("/login");
        }
    }, [logout, navigate, removeCookie]);

    return () => setLogout(true);
};

export const hyphenate = (string) => {
    return string.trim().replace(" ", "-");
};

export const useAPI = (apiCallback, varArray = [], lazy = false, auth = true) => {
    const [cookies] = useCookies(TOKEN_COOKIE);
    const navigate = useNavigate();

    const [callAPI, setCallAPI] = useState(!lazy);
    const [data, setData] = useState(null);
    const [loading, setLoading] = useState(!lazy);
    const [error, setError] = useState(null);

    const callback = () => {
        setLoading(true);
        setCallAPI(true);
    };

    const authHeaders = {
        headers: {
            authorization: cookies[TOKEN_COOKIE],
        },
    };

    useEffect(() => {
        if (callAPI) {
            apiCallback
                .apply(null, auth ? [authHeaders, ...varArray] : [...varArray])
                .then((response) => {
                    setData(response.data);
                })
                .catch((error) => {
                    if (unauthorisedError(error)) {
                        navigate("/login");
                    } else {
                        setError(error.response.data.error);
                    }
                })
                .finally(() => {
                    setLoading(false);
                    setCallAPI(false);
                });
        }
    }, [apiCallback, auth, callAPI, navigate, varArray]);

    return { data, loading, error, callback };
};

export const getSelectOptionsFromArray = (valArray = [], labelArray = []) => {
    return valArray.length === labelArray.length
        ? valArray.map((val, i) => {
              return { value: val, label: labelArray[i] };
          })
        : valArray.map((val) => {
              return { value: val, label: val };
          });
};

export const getAssignableSystemRoleOptions = (currentUserSystemRole) => {
    const systemRolesOptions = getSelectOptionsFromArray(SYSTEM_ROLES);
    switch (currentUserSystemRole) {
        case ADMIN:
            return systemRolesOptions;
        case MANAGER_SR:
            return systemRolesOptions.filter((option) => {
                return option.label !== ADMIN && option;
            });
        default:
            return systemRolesOptions.filter((option) => {
                return option.label !== ADMIN && option.label !== MANAGER_SR && option;
            });
    }
};

export const isValidString = (val, min = 0, max = 100) => {
    return !!val && val.length >= min && val.length <= max;
};

export const isValidEmail = (val) => {
    return /\S+@\S+\.\S+/.test(val);
};

export const isValidUser = (user) => {
    return (
        !!user.firstName &&
        isValidString(user.firstName, 1) &&
        !!user.lastName &&
        isValidString(user.lastName, 1) &&
        !!user.email &&
        isValidEmail(user.email) &&
        (!!user.password ? isValidString(user.password, 10, 16) : true)
    );
};

export const useIsAdminOrManager = () => {
    const [currentUser, setShowToast] = useOutletContext();
    const navigate = useNavigate();

    useEffect(() => {
        if (
            currentUser &&
            currentUser.system_role !== ADMIN &&
            currentUser.system_role !== MANAGER_SR
        ) {
            setShowToast({ error: "Unauthorised route" });
            navigate("/");
        }
    });
};

const unauthorisedError = (error) => error.response.status === 401 || error.response.status === 403;

export const formatDate = (_date) => {
    if (!_date) return "N/A";
    let newDate = new Date(_date);
    newDate.toLocaleDateString();
    return newDate.toLocaleDateString();
};

export const formatSkillLevel = (_skillLevel) => {
    if (_skillLevel === 5) {
        return EXPERT;
    } else if (_skillLevel === 4) {
        return HIGH;
    } else if (_skillLevel === 3) {
        return INTERMEDIATE;
    } else if (_skillLevel === 2) {
        return BASIC;
    } else {
        return NONE;
    }
};

export const convertSkillLevelName = (_skillLevel) => {
    if (_skillLevel === "Expert") {
        return 5;
    } else if (_skillLevel === "High") {
        return 4;
    } else if (_skillLevel === "Intermediate") {
        return 3;
    } else if (_skillLevel === "Basic") {
        return 2;
    } else {
        return 1;
    }
};

export const expiryDateNullFormatter = (_expiryDate) => {
    if (_expiryDate === null) {
        return "N/A";
    } else {
        return _expiryDate;
    }
};
