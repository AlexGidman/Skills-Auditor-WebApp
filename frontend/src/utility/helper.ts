import { useEffect, useState } from "react";
import { useOutletContext, useNavigate } from "react-router-dom";
import {
    ADMIN,
    BASIC,
    EXPERT,
    HIGH,
    NONE,
    INTERMEDIATE,
    MANAGER_SR,
    systemRoleS,
    TOKEN_COOKIE,
    User,
} from "./types";
import { useCookies } from "react-cookie";
import { AxiosResponse, AxiosError } from "axios";

export const useLogout = () => {
    const [, , removeCookie] = useCookies([TOKEN_COOKIE]); // TODO: Check this works!
    const navigate = useNavigate();
    const [logout, setLogout] = useState(false);

    useEffect(() => {
        if (logout) {
            setLogout(false);
            removeCookie(TOKEN_COOKIE);
            navigate("/login");
        }
    }, [logout, navigate, removeCookie]);

    return () => setLogout(true);
};

export const hyphenate = (str: string) => {
    return str.trim().replace(" ", "-");
};

export function useAPI<T>(
    apiCallback: Function,
    varArray: unknown[] = [],
    lazy = false,
    auth = true,
): { data: T | null; loading: boolean; error: Error | null; callback: CallableFunction } {
    const [cookies] = useCookies([TOKEN_COOKIE]); // TODO: Check this works!
    const navigate = useNavigate();

    const [callAPI, setCallAPI] = useState<boolean>(!lazy);
    const [data, setData] = useState<T | null>(null);
    const [loading, setLoading] = useState<boolean>(!lazy);
    const [error, setError] = useState<Error | null>(null);

    const callback = () => {
        setLoading(true);
        setCallAPI(true);
    };

    useEffect(() => {
        if (callAPI) {
            const authHeaders = {
                headers: {
                    authorization: cookies[TOKEN_COOKIE],
                },
            };

            apiCallback
                .apply(null, auth ? [authHeaders, ...varArray] : [...varArray])
                ?.then((response: AxiosResponse) => {
                    setData(response.data);
                })
                .catch((error: AxiosError<{ error: Error }>) => {
                    if (unauthorisedError(error)) {
                        navigate("/login");
                    } else {
                        setError(error?.response?.data?.error || null);
                    }
                })
                .finally(() => {
                    setLoading(false);
                    setCallAPI(false);
                });
        }
    }, [apiCallback, auth, callAPI, cookies, navigate, varArray]);

    return { data, loading, error, callback };
}

export const getSelectOptionsFromArray = (valArray: string[] = [], labelArray: string[] = []) => {
    return valArray.length === labelArray.length
        ? valArray.map((val, i) => {
              return { value: val, label: labelArray[i] };
          })
        : valArray.map((val) => {
              return { value: val, label: val };
          });
};

export const getAssignableSystemRoleOptions = (currentUserSystemRole: string) => {
    // TODO implement proper type for system role
    const systemRolesOptions = getSelectOptionsFromArray(systemRoleS);
    switch (currentUserSystemRole) {
        case ADMIN:
            return systemRolesOptions;
        case MANAGER_SR:
            return systemRolesOptions.filter((option) => {
                return option.label !== ADMIN && option;
            });
        default:
            return systemRolesOptions.filter((option) => {
                return option.label !== ADMIN && option.label !== MANAGER_SR && option;
            });
    }
};

export const isValidString = (val: string, min = 0, max = 100) => {
    return !!val && val.length >= min && val.length <= max;
};

export const isValidEmail = (val: string) => {
    return /\S+@\S+\.\S+/.test(val);
};

export const isValidUser = (user: User) => {
    return (
        !!user.firstName &&
        isValidString(user.firstName, 1) &&
        !!user.lastName &&
        isValidString(user.lastName, 1) &&
        !!user.email &&
        isValidEmail(user.email) &&
        (!!user.password ? isValidString(user.password, 10, 16) : true)
    );
};

export const useIsAdminOrManager = () => {
    // @ts-ignore TODO fix type for AppOutletContext here
    const [currentUser, setShowToast] = useOutletContext();
    const navigate = useNavigate();

    useEffect(() => {
        if (
            currentUser &&
            currentUser.systemRole !== ADMIN &&
            currentUser.systemRole !== MANAGER_SR
        ) {
            setShowToast({ error: "Unauthorised route" });
            navigate("/");
        }
    });
};

const unauthorisedError = (error: AxiosError): boolean => {
    return (
        !!!error?.response?.status ||
        error?.response?.status === 401 ||
        error?.response?.status === 403
    );
};

export const formatDate = (_date: string) => {
    if (!_date || _date === "0000-00-00") return "N/A";
    let newDate = new Date(_date);
    newDate.toLocaleDateString();
    return newDate.toLocaleDateString();
};

export const formatSkillLevel = (_skillLevel: string) => {
    // TODO implement proper type for skill level
    if (_skillLevel === "5") {
        return EXPERT;
    } else if (_skillLevel === "4") {
        return HIGH;
    } else if (_skillLevel === "3") {
        return INTERMEDIATE;
    } else if (_skillLevel === "2") {
        return BASIC;
    } else {
        return NONE;
    }
};

export const convertSkillLevelName = (_skillLevel: string) => {
    // TODO implement proper type for skill level
    if (_skillLevel === "Expert") {
        return "5";
    } else if (_skillLevel === "High") {
        return "4";
    } else if (_skillLevel === "Intermediate") {
        return "3";
    } else if (_skillLevel === "Basic") {
        return "2";
    } else {
        return "1";
    }
};
