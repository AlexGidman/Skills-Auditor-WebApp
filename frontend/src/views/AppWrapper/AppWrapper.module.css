.content {
    overflow-x: hidden;
    overflow-y: auto;
    padding: var(--s2);
    margin-left: calc(var(--s1) * 17);
    text-align: center;
}
.hamburger {
    display: none;
}

.header {
    display: flex;
    align-items: center;
    justify-content: flex-end;
    min-height: var(--s6);
    padding: var(--s1);
    color: var(--c-secondary);
    background-color: var(--c-primary);
}

@media (max-width: 900px) {
    .header {
        justify-content: space-between;
    }
    .nav {
        display: none;
    }
    .hamburger {
        display: inline;
    }
    .content {
        margin-left: 0;
    }
}
