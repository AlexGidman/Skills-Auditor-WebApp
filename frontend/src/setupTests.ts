// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom";
import * as apiRequests from "./utility/apiRequests";
import * as reactCookie from "react-cookie";

import { MANAGER_JR, ADMIN, User } from "./utility/types";

jest.mock("uuid", () => {
    return {
        ...jest.requireActual("uuid"),
        v4: jest.fn().mockReturnValue("uuid"),
    };
});

jest.mock("react-cookie", () => {
    return {
        ...jest.requireActual("react-cookie"),
        useCookies: jest.fn(),
    };
});

export const mockApiRequests = {
    getAllCategories: jest.spyOn(apiRequests, "getAllCategories"),
    getCategory: jest.spyOn(apiRequests, "getCategory"),
    updateCategory: jest.spyOn(apiRequests, "updateCategory"),
    getUserDetails: jest.spyOn(apiRequests, "getUserDetails"),
    updateUserDetails: jest.spyOn(apiRequests, "updateUserDetails"),
    getSkill: jest.spyOn(apiRequests, "getSkill"),
    updateSkill: jest.spyOn(apiRequests, "updateSkill"),
    getAllSkills: jest.spyOn(apiRequests, "getAllSkills"),
    addUser: jest.spyOn(apiRequests, "addUser"),
    addDirectReport: jest.spyOn(apiRequests, "addDirectReport"),
    getUserFromToken: jest.spyOn(apiRequests, "getUserFromToken"),
    getAllDirectReports: jest.spyOn(apiRequests, "getAllDirectReports"),
    getAllStaffSkills: jest.spyOn(apiRequests, "getAllStaffSkills"),
    deleteUser: jest.spyOn(apiRequests, "deleteUser"),
    getStaffSkill: jest.spyOn(apiRequests, "getStaffSkill"),
    updateStaffSkill: jest.spyOn(apiRequests, "updateStaffSkill"),
    postLogin: jest.spyOn(apiRequests, "postLogin"),
    addSkill: jest.spyOn(apiRequests, "addSkill"),
    addCategory: jest.spyOn(apiRequests, "addCategory"),
    deleteStaffSkill: jest.spyOn(apiRequests, "deleteStaffSkill"),
    addUserSkill: jest.spyOn(apiRequests, "addUserSkill"),
};

export const mockAdminUser = {
    id: "1",
    firstName: "Admin",
    lastName: "User",
    email: "Admin@email.com",
    jobRole: MANAGER_JR,
    systemRole: ADMIN,
};

export const mockError = { response: { data: { error: new Error("Oops") }, status: 400 } };

beforeAll(() => {
    console.log = jest.fn();
    console.error = jest.fn();
});

beforeEach(() => {
    jest.spyOn(reactCookie, "useCookies").mockImplementation(() => [
        { token: "Bearer token" },
        jest.fn(),
        jest.fn(),
    ]);
    // @ts-ignore
    mockApiRequests.getUserFromToken.mockImplementation(() =>
        Promise.resolve({ data: mockAdminUser }),
    );
});

afterEach(() => {
    jest.resetAllMocks();
});
