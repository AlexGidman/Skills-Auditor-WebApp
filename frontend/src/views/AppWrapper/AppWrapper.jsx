import React, { useEffect, useState } from "react";
import { Outlet } from "react-router-dom";
import { toast } from "react-toastify";
import { useCookies } from "react-cookie";

import { LoadingPlaceholder } from "../../components/LoadingPlaceholder/LoadingPlaceholder";
import { ErrorMessage } from "../../components/ErrorMessage/ErrorMessage";
import { NavBar, HamburgerNavBar } from "../../components/NavBar/NavBar";
import { UserPill } from "../../components/UserPill/UserPill";
import { useLogout, useAPI } from "../../utility/helper";
import { getUserFromToken } from "../../utility/apiRequests";
import { TOKEN_COOKIE } from "../../utility/types";
import styles from "./AppWrapper.module.css";

export const AppWrapper = () => {
    const [cookies] = useCookies(TOKEN_COOKIE);
    const processLogout = useLogout();
    const [showToast, setShowToast] = useState();

    const { data: currentUser, error, callback } = useAPI(getUserFromToken, [], true);

    useEffect(() => {
        cookies[TOKEN_COOKIE] && callback();
    }, [cookies]);

    useEffect(() => {
        if (!cookies[TOKEN_COOKIE]) processLogout();
    }, [cookies, processLogout]);

    useEffect(() => {
        if (showToast) {
            showToast.success && toast.success(showToast.success);
            showToast.error && toast.error(showToast.error);
            setShowToast(undefined);
        }
    }, [showToast]);

    if (error) return <ErrorMessage error={error} />;
    if (currentUser)
        return (
            <>
                <header className={styles.header}>
                    <HamburgerNavBar
                        systemRole={currentUser.system_role}
                        className={styles.hamburger}
                    />
                    <UserPill
                        systemRole={currentUser.system_role}
                        email={currentUser.email}
                        className={styles.userPill}
                    />
                </header>
                <NavBar systemRole={currentUser.system_role} className={styles.nav} />
                <div className={styles.content}>
                    <Outlet context={[currentUser, setShowToast]} />
                </div>
            </>
        );
    return <LoadingPlaceholder />;
};
