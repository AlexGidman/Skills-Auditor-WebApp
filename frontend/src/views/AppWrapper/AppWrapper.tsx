import React, { useEffect, useState } from "react";
import { Outlet } from "react-router-dom";
import { toast } from "react-toastify";
import { useCookies } from "react-cookie";

import { LoadingPlaceholder } from "../../components/LoadingPlaceholder/LoadingPlaceholder";
import { ErrorMessage } from "../../components/ErrorMessage/ErrorMessage";
import { NavBar, HamburgerNavBar } from "../../components/NavBar/NavBar";
import { UserPill } from "../../components/UserPill/UserPill";
import { useLogout, useAPI } from "../../utility/helper";
import { getUserFromToken } from "../../utility/apiRequests";
import { TOKEN_COOKIE, User } from "../../utility/types";
import styles from "./AppWrapper.module.css";

export const AppWrapper = () => {
    const [cookies] = useCookies([TOKEN_COOKIE]);
    const processLogout = useLogout();
    const [showToast, setShowToast] = useState<{ error?: string; success?: string }>();

    const { data: currentUser, error, callback } = useAPI<User>(getUserFromToken, [], true);

    useEffect(() => {
        cookies[TOKEN_COOKIE] && callback();
    }, [cookies]);

    useEffect(() => {
        if (!cookies[TOKEN_COOKIE]) processLogout();
    }, [cookies, processLogout]);

    useEffect(() => {
        if (showToast) {
            showToast.success && toast.success(showToast.success);
            showToast.error && toast.error(showToast.error);
            setShowToast(undefined);
        }
    }, [showToast]);

    if (error) return <ErrorMessage error={error} />;
    if (currentUser)
        return (
            <>
                <header className={styles.header}>
                    <HamburgerNavBar
                        systemRole={currentUser.systemRole}
                        className={styles.hamburger}
                    />
                    <UserPill
                        systemRole={currentUser.systemRole}
                        email={currentUser.email}
                        className={styles.userPill}
                    />
                </header>
                <NavBar systemRole={currentUser.systemRole} className={styles.nav} />
                <div className={styles.content}>
                    <Outlet context={[currentUser, setShowToast]} />
                </div>
            </>
        );
    return <LoadingPlaceholder />;
};
