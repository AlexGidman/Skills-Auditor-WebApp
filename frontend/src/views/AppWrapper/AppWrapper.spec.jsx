import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { mockApiRequests, mockAdminUser } from "../../setupTests";
import * as rrd from "react-router-dom";
import * as reactCookie from "react-cookie";
import * as helper from "../../utility/helper";
import { wait } from "@testing-library/user-event/dist/utils";

import { AppWrapper } from "./AppWrapper";

jest.mock("react-router-dom", () => {
    return {
        ...jest.requireActual("react-router-dom"),
        useOutletContext: jest.fn(),
        useNavigate: jest.fn(),
    };
});

describe("AppWrapper", () => {
    it("renders correctly", async () => {
        jest.spyOn(reactCookie, "useCookies").mockImplementation(() => [
            { token: "Bearer token" },
            jest.fn(),
            jest.fn(),
        ]);

        mockApiRequests.getUserFromToken.mockImplementation(() =>
            Promise.resolve({ data: mockAdminUser }),
        );

        const mockNavigate = jest.fn();
        jest.spyOn(rrd, "useNavigate").mockImplementation(() => mockNavigate);
        const { container } = render(
            <MemoryRouter>
                <AppWrapper />
            </MemoryRouter>,
        );
        await wait(200);
        expect(screen.getAllByRole("navigation").length).toBe(2);
        expect(screen.getAllByRole("link").length).toBe(16);
        expect(
            screen.getByRole("status", { name: "You are logged in as an admin" }),
        ).toBeInTheDocument();
        expect(
            screen.getByRole("status", { name: `${mockAdminUser.email} is logged in` }),
        ).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });
    it("should log user out if user is not authenticated", async () => {
        const spyLogout = jest.spyOn(helper, "useLogout");
        jest.spyOn(reactCookie, "useCookies").mockImplementation(() => [
            { token: null },
            jest.fn(),
            jest.fn(),
        ]);
        jest.spyOn(rrd, "useNavigate").mockImplementation(() => jest.fn());
        expect(spyLogout).not.toHaveBeenCalled();
        render(
            <MemoryRouter>
                <AppWrapper />
            </MemoryRouter>,
        );
        await wait(200);
        expect(spyLogout).toHaveBeenCalled();
    });
});
