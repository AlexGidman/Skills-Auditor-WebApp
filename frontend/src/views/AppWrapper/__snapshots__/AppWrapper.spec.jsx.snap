// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppWrapper renders correctly 1`] = `
<div>
  <header
    class="header"
  >
    <nav
      class="hamburger"
    >
      <button
        aria-label="Navigation"
        class="button"
      >
        <img
          alt="menu"
          aria-hidden="true"
          class="hamburgerIcon"
          src="hamburger.svg"
        />
      </button>
      <ul
        class="hamburgerLinks hidden"
      >
        <li
          class="listItem"
        >
          <a
            class="component"
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="listItem"
        >
          <a
            class="component"
            href="/mydetails"
          >
            My Details
          </a>
        </li>
        <li
          class="listItem"
        >
          <a
            class="component"
            href="/myskills"
          >
            My Skills
          </a>
        </li>
        <li
          class="listItem"
        >
          <a
            class="component"
            href="/viewstaff"
          >
            View Staff
          </a>
        </li>
        <li
          class="listItem"
        >
          <a
            class="component"
            href="/addstaff"
          >
            Add Staff
          </a>
        </li>
        <li
          class="listItem"
        >
          <a
            class="component"
            href="/viewskills"
          >
            View Skills
          </a>
        </li>
        <li
          class="listItem"
        >
          <a
            class="component"
            href="/addskill"
          >
            Add Skill
          </a>
        </li>
        <li
          class="listItem"
        >
          <a
            class="component"
            href="/viewcategories"
          >
            View Categories
          </a>
        </li>
        <li
          class="listItem"
        >
          <button
            class="component signoutButton"
            name="Sign Out"
            path="/login"
          >
            Sign out
          </button>
        </li>
      </ul>
    </nav>
    <div
      class="component userPill"
    >
      <span
        aria-label="You are logged in as an admin"
        class="admin"
        role="status"
      >
        ADMIN
      </span>
      <span
        aria-label="Admin@email.com is logged in"
        class="pill"
        role="status"
      >
        Admin@email.com
      </span>
    </div>
  </header>
  <nav
    class="component nav"
  >
    <img
      alt="logo"
      aria-hidden="true"
      class="logo"
      src="logo.svg"
    />
    <ul>
      <li
        class="listItem"
      >
        <a
          class="component link"
          href="/"
        >
          Home
        </a>
      </li>
      <li
        class="listItem"
      >
        <a
          class="component link"
          href="/mydetails"
        >
          My Details
        </a>
      </li>
      <li
        class="listItem"
      >
        <a
          class="component link"
          href="/myskills"
        >
          My Skills
        </a>
      </li>
      <li
        class="listItem"
      >
        <a
          class="component link"
          href="/viewstaff"
        >
          View Staff
        </a>
      </li>
      <li
        class="listItem"
      >
        <a
          class="component link"
          href="/addstaff"
        >
          Add Staff
        </a>
      </li>
      <li
        class="listItem"
      >
        <a
          class="component link"
          href="/viewskills"
        >
          View Skills
        </a>
      </li>
      <li
        class="listItem"
      >
        <a
          class="component link"
          href="/addskill"
        >
          Add Skill
        </a>
      </li>
      <li
        class="listItem"
      >
        <a
          class="component link"
          href="/viewcategories"
        >
          View Categories
        </a>
      </li>
      <li
        class="listItem"
      >
        <button
          class="component signoutButton"
          name="Sign Out"
          path="/login"
        >
          Sign out
        </button>
      </li>
    </ul>
  </nav>
  <div
    class="content"
  />
</div>
`;
